### ISSUE TITLE
[Help wanted] Fine tune pre-trained muss(mbart-large-cc25) with mined paraphrases (si_LK)

### ISSUE THREAD

**Issue by NomadXD** [facebookresearch/muss#15]
For my final year research project, I'm using this approach as my baseline and trying to do text simplification for Sinhala language (native language used in Sri Lanka). I don't have enough infrastructure to run the `cc_net` pipeline to crawl language data but I used several already existing sources (15.7 million plain sentences) and mined them to get  paraphrases. Now as the next stage I'm trying to fine tune mbart using the mined paraphrases. I went through the `train_model.py` and `train_paper_models.py` and implemented a similar script for Sinhala by making changes to the related methods.

```py
from muss.fairseq.main import fairseq_train_and_evaluate_with_parametrization
from muss.mining.training import get_mbart_kwargs


sin15M = 'sin15M'
kwargs = get_mbart_kwargs(dataset=sin15M, language='si', use_access=False)
kwargs['train_kwargs']['ngpus'] = 1  # Set this from 8 to 1 for local training
kwargs['train_kwargs']['max_tokens'] = 512  # Lower this number to prevent OOM
result = fairseq_train_and_evaluate_with_parametrization(**kwargs)
``` 
I tried to run this for a small sample of 6000 train sentences, 750 test and valid sentences in a nvidia tesla t4 with 16GB of GPU memory. I get OOM issues when I tried the `kwargs['train_kwargs']['max_tokens']` with values 512,256,64,32,16,8. With 4 it says no dataset found. Any idea what could go wrong here ???

PS - I did not change anything internally related to the models. Only changed the helper methods to get datasets.

**Comment by louismartin** [comment]
Hi,
I'm not sure what is the problem here.
Can you paste the output of the script along with the error?

**Comment by NomadXD** [comment]
@louismartin Thanks a lot for the response and sorry for the late reply. I fixed the OOM issue by tweaking the hyper-parameters like batch size, max_tokens etc and added `--memory-efficient-fp16`. So after fine tuning with 6000 mined sentences, the model predict mostly paraphrases and very few simplifications. 

Now I'm trying to get a larger parallel mined paraphrase dataset by running the `scripts/mine_sequences.py`. But I couldn't run it for 15.7 million unparalleled sentences in GCP compute engine instance with 16 cores, 102 memory and nvidia tesla t4 gpu. I think the issue is with generating LASER embeddings. (No issue with the memory it's 60% in the maximum case but CPU is 100% and GPU is around 90%). Is there any specific configuration that I could use to run the mining in a low resource environment ? (like the GCP environment that I mentioned). Or else what do you think about the minimum infrastructure required to mine 15.7 million sentences for paraphrases ? 

PS - For the LASER embedding generation I changed the `max_tokens` to 1024 instead of 3000. Also is there any effect from `slurm_array_parallelism` when I run locally ?

Thanks in advance.

```py
get_embeddings = lambda sentences: get_laser_embeddings(
    sentences, max_tokens=1024, language=language, n_encoding_jobs=16
)  # noqa: E731
```

**Comment by louismartin** [comment]
Hi @NomadXD 
Thanks for the questions.
Overall the mining process is very resource intensive and was run in parallel on hundreds of instances.
Maybe try with a smaller amount of sentences and increase iteratively.

You can maybe remove the `with mute():` statement [here](https://github.com/facebookresearch/muss/blob/ed70446abdbebce4513123c376e40f5ff05290a0/muss/laser.py#L45) to get more information about laser embeddings.
The mining stores many intermediary files so maybe that can be a problem as well.


I don't think slurm_array_parallelism has any effect.

**Comment by chirico85** [comment]
@NomadXD and @louismartin 
I am also using mbart-large-cc25 with mined german paraphrases. After training, inference provides only dots and commas. Did you experience similar issues in your case?
I must say, that I had to prune pretrained mbart before training due to OOM issues on 1 gpu. 
I did that following these scripts #https://github.com/pytorch/fairseq/issues/2120#issuecomment-647429120 and #https://github.com/pytorch/fairseq/issues/2120#issuecomment-661622230

I must also admit that before finetuning mbart I mistakenly used the same mined german paraphrases to finetune bart (without prunning). Inferencing worked and provided some "ok" results. However, that was better than just dots and commas.

During finetuning with bart and mbart I had no proper german evaluation dataset for fairseq_evaluate_and_save (just used the provided asset dataset in english). As far as I understood that should not affect the training results and is only needed to get a score on prediction, right?

Hope you can help me!

**Comment by louismartin** [comment]
Hi @chirico85 

I haven't experienced only dots and commas. It might be due to pruning mbart. Can you maybe prune other layers like some layers in the middle? Or one layer out of two spread out in the whole model?

Yes it seems that BART know some other languages out of the box.

I don't think having no german evaluation dataset is problematic, it's only used for evaluation (maybe double check that is doesn't cause problems for early stopping but I don't think it's the case).

**Comment by chirico85** [comment]
Hello @louismartin,
Again, thanks a lot for your fast reply. I tried getting some help on the fairseq repo issues section where they discuss the approach of prunning mbart, too. No answer yet.

I did not try out yet prunning other layers like some layers in the middle. which effect do you expect? 
As far as I understand, the approach just incorporates reducing the dictionary size of pre-trained mbart by maintaining the words of my corpus dictionary and the original indices.

**Comment by chirico85** [comment]
Hello @louismartin,
It works now. 
Unfortunately I cannot exactly say why it was not working. I think the error must be in the wrong format of dict.txt after preprocessing with bpe and later inferencing.  Somewhat did not match (thats why I got many unks). Prunning was not the reason. 
However, I ran all the preprocessing and training step by step as described in some of the comments in #https://github.com/pytorch/fairseq/issues/2120 and [here ](https://github.com/pytorch/fairseq/tree/main/examples/multilingual#training).

Unfortunately, validation score did not decrease after some few epochs (what can I assume from this?), while loss was still decreasing even after 40k steps.

Regards
